File: 0022_generate_parenthesis.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        dp = [[] if i else [""] for i in range(n + 1)]
        for i in range(1, n + 1):
            for j in range(i):
                dp[i] += ['(' + x + ')' + y for x in dp[j] for y in dp[i - j - 1]]
        return dp[n]

# Alternative solution

class Solution1:
    def generateParenthesis(self, n: int) -> List[str]:
        output = []
        path = ""
        self.dfs(n, output, path)
        return output
        
    def dfs(self, n, output, path):
        if len(path) == 2 * n:
            output.append(path)
            return
        for i in range(n):
            if path.count('(') < n:
                self.dfs(n, output, path + '(')
            if path.count('(') > path.count(')'):
                self.dfs(n, output, path + ')')
            return

File: 0001-0500/0022_generateParenthesis.java
class Solution {
    public List<String> generateParenthesis(int n) {
        if (n == 0) {
            List<String> base = new ArrayList<>();
            base.add("");
            return base;
        }

        List<String> output = new ArrayList<>();

        for (int i = 0; i < n; i++) {
            for (String left: generateParenthesis(i)) {
                for (String right: generateParenthesis(n - i - 1)) {
                    output.add("(" + left + ")" + right);
                }
            }
        }

        return output;
    }
}
